cpp/src/arrow/array.cc

```diff
@@ -125,6 +126,26 @@ std::string Array::ToString() const {
   return ss.str();
 }
 
+template <>
+Status Array::castTyped<DictionaryType>(std::shared_ptr<Array>* out) {
+  MemoryPool* pool = default_memory_pool();
+  RETURN_NOT_OK(MakeDictionaryArray(pool, *this, out));
+  return Status::OK();
+};
+
+Status Array::cast(const std::shared_ptr<DataType>& type, std::shared_ptr<Array>* out) {
+  switch (type->id()) {
+    case Type::DICTIONARY:
+      RETURN_NOT_OK(castTyped<DictionaryType>(out));
+      return Status::OK();
+    default:
+      std::stringstream ss;
+      ss << "Unhandled conversion type from " << this->type()->ToString();
+      ss << " to " << type->ToString();
+      return Status::NotImplemented(ss.str());
+  }
+};
```

Comment:
Can you instead create a function `Status EncodeDictionary(const Array& array, MemoryPool* pool, std::shared_ptr<Array>* out);`. This will be a little more manageable for the scope of this issue (and we can reuse it later when expanding casting functionality)